app/controllers/shopify_app/callback_controller.rb

```diff
@@ -26,6 +31,29 @@ def callback
 
     private
 
+    def jwt_callback
+      if valid_jwt_auth?
+        create_user_session_from_jwt_callback
+        head(:ok)
+      else
+        head(:unauthorized)
+      end
+    end
+
+    def valid_jwt_auth?
+      auth_hash && jwt_shopify_domain == shop_name && jwt_shopify_user_id == associated_user_id
+    end
+
+    def create_user_session_from_jwt_callback
```

Comment:
The relationship between this and the existing `set_shopify_session` is a little confusing to me. If we need the behaviours to be different, I wonder if we should factor out some of the common elements (like the actual `ShopifyAPI::Session.new` call) to highlight why these need to be different. Or, if they don't need to have different logic, can we just use `set_shopify_session` and not have this method?